File: C++/TwoSum.cpp
/*
    Problem:
    :Given an array of integers nums and an integer target,
       return indices of the two numbers such that they add up to target.

    Constraints:
    :None

    Pseudo:
    :currentSum = 0
     for num in list
      for num2 in list
         currentSum = num + num2
         if currentSum is == target
            reutn num and num 2

    Experiments:
    :Experiment-1:
    :Iterating through a vector<int> and returning each number

    Reductions:
    :Reduction-1:
    :Return every possible sum between two numbers
*/

#include <vector>
#include <iostream>

// Solution: 15min. 21sec.
/*
    Problems:
    :a self-sum isn't valid
*/
std::vector<int> twoSum(std::vector<int> &nums, int target)
{
    std::vector<int> indices = {};

    int currentSum = 0;

    for (int i = 0; i < nums.size(); i++)
    {
        for (int j = 0; j < nums.size(); j++)
        {
            currentSum = nums[i] + nums[j];

            // a self-sum isn't valid
            if (j == i)
            {
                continue;
            }

            if (currentSum == target)
            {
                std::cout << nums[i] << " " << nums[j] << " " << i << " " << j << " " << target << std::endl;
                indices.push_back(i);
                indices.push_back(j);

                return indices;
            }
        }
    }

    return indices;
}

int main()
{
    std::vector<int> input = {3, 2, 4};

    std::vector<int> indices = twoSum(input, 6);

    std::cout << indices[0] << " " << indices[1];
}

// Reduction-1:
// #include <vector>
// #include <iostream>
// std::vector<int> convertNumsToSums(std::vector<int> nums)
// {
//     int currentSum;

//     std::vector<int> sums = {};

//     for (int &num1 : nums)
//     {
//         for (int &num2 : nums)
//         {
//             currentSum = num1 + num2;

//             sums.push_back(currentSum);
//         }
//     }

//     return sums;
// }

// int main()
// {
//     std::vector<int> nums = {8, 3, 5, 3, 1, 4};

//     std::vector<int> sums = convertNumsToSums(nums);

//     for (int numsIndex = 0; numsIndex < nums.size(); numsIndex++)
//     {
//         std::cout << "Sums of " << nums[numsIndex] << ": ";

//         for (int sumsIndex = numsIndex * nums.size(); sumsIndex < (numsIndex * nums.size()) + nums.size(); sumsIndex++)
//         {
//             std::cout << sums[sumsIndex] << " ";
//         }

//         std::cout << std::endl;
//     }

//     return 0;
// }

// Experiment-1:
// #include <vector>
// #include <iostream>

// int main()
// {
//     std::vector<int> numsList = {1, 2, 6, 7, 2, 9};

//     for (int i = 0; i < numsList.size(); i++)
//     {
//         std::cout << "Index " << i << ": " << numsList.at(i) << std::endl;
//     }

//     std::cout << "Finished!";

//     return 0;
// }

File: Java/TwoSum.java
import java.util.Arrays;

/**
 * Problem:
 * :Given an array of integers nums and an integer target,
 * return indices of the two numbers such that they add
 * up to target.
 * 
 * Constraints:
 * :Each input has one solution
 * :You may noy use the same element twice
 * :The answer can be returned in any order
 * 
 * Want:
 * :The indices that sum to the target as an array
 * 
 * Pseudo:
 * :Iterate through the array
 * :subtract each number from the target
 * :iterate through the array again looking for the difference
 * :if the difference is not found, move to the next index
 * :repeat until a matching difference is found
 */
public class TwoSum {
    public static void main(String[] args) {
        int[] data1 = { 2, 7, 11, 15 };
        int[] data2 = { 3, 2, 4 };

        int target1 = 9;
        int target2 = 6;

        System.out.println(Arrays.toString(twoSum(data1, target1)));
        System.out.println(Arrays.toString(twoSum(data2, target2)));
    }

    /**
     * Implementation of the solution algorithm.
     * Solved: 03.28.2023
     * 
     * Takes an array of integers and a target integer
     * and return the array of the two indices that sum to the given
     * target.
     * 
     * @param nums the array of integers that include the two addends
     * @param target the target sum
     * @return the two addends of the sum as an integer array
     */
    public static int[] twoSum(int[] nums, int target) {
        int[] indices = new int[2];
        boolean indicesFound = false;

        // Iterate through the array and calculate a possible difference
        //      that will sum to the target
        for (int i = 0; i < nums.length; i++) {
            int difference = target - nums[i];

            // because this means the current addend is greater than the target
            if (difference < 0)
                continue;

            // Iterate through the array beginning at the next index(because the 
            //      previoud indices have already been checked) and look for a 
            //      number that equals the difference; if one us found, add the
            //      indices to the returning array.
            for (int j = i + 1; j < nums.length; j++) {
                if (nums[j] == difference) {
                    indicesFound = true;

                    indices[0] = i;
                    indices[1] = j;

                    break;
                }
            }

            if (indicesFound)
                break;
        }

        return indices;
    }
}